lib/web/assistant.go

```diff
@@ -19,46 +19,785 @@
+	// messageKindAssistantPartialMessage is the type of Assist message that contains the assistant partial message.
+	messageKindAssistantPartialMessage assistantMessageType = "CHAT_PARTIAL_MESSAGE_ASSISTANT"
+	// messageKindAssistantPartialFinalize is the type of Assist message that ends the partial message stream.
+	messageKindAssistantPartialFinalize assistantMessageType = "CHAT_PARTIAL_MESSAGE_ASSISTANT_FINALIZE"
+	// messageKindSystemMessage is the type of Assist message that contains the system message.
+	messageKindSystemMessage assistantMessageType = "CHAT_MESSAGE_SYSTEM"
+)
+
+// createAssistantConversationResponse is a response for POST /webapi/assistant/conversations.
+type createdAssistantConversationResponse struct {
+	// ID is a conversation ID.
+	ID string `json:"id"`
+}
+
+// createAssistantConversation is a handler for POST /webapi/assistant/conversations.
 func (h *Handler) createAssistantConversation(_ http.ResponseWriter, r *http.Request,
 	_ httprouter.Params, sctx *SessionContext,
 ) (any, error) {
-	return nil, trace.NotImplemented("not implemented")
+	authClient, err := sctx.GetClient()
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	if err := checkAssistEnabled(authClient, r.Context()); err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	req := &assist.CreateAssistantConversationRequest{
+		CreatedTime: timestamppb.New(h.clock.Now().UTC()),
+		Username:    sctx.GetUser(),
+	}
+
+	resp, err := authClient.CreateAssistantConversation(r.Context(), req)
+	if err != nil {
+		return nil, err
+	}
+
+	return &createdAssistantConversationResponse{
+		ID: resp.Id,
+	}, nil
 }
 
+// assistantMessage is an assistant message that is sent to the client.
+type assistantMessage struct {
+	// Type is a type of the message.
+	Type assistantMessageType `json:"type"`
+	// CreatedTime is a time when the message was created in RFC3339 format.
+	CreatedTime string `json:"created_time"`
+	// Payload is a message payload in JSON format.
+	Payload string `json:"payload"`
+}
+
+// getAssistantConversation is a handler for GET /webapi/assistant/conversations/:conversation_id.
 func (h *Handler) getAssistantConversationByID(_ http.ResponseWriter, r *http.Request,
 	p httprouter.Params, sctx *SessionContext,
 ) (any, error) {
-	return nil, trace.NotImplemented("not implemented")
+	authClient, err := sctx.GetClient()
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	if err := checkAssistEnabled(authClient, r.Context()); err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	conversationID := p.ByName("conversation_id")
+
+	resp, err := authClient.GetAssistantMessages(r.Context(), &assist.GetAssistantMessagesRequest{
+		ConversationId: conversationID,
+		Username:       sctx.GetUser(),
+	})
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	return conversationResponse(resp), nil
+}
+
+// conversationResponse creates a response for GET conversation response.
+func conversationResponse(resp *assist.GetAssistantMessagesResponse) any {
```

Comment:
should this sig have any in it? It seems like throwing away more information than we have to and makes this function harder to grasp at a glance when I e.g. hover over it in my IDE?

Reply:
I'll fix that.